### ISSUE TITLE
Strange VO verbosity when opening Keyboard Shortcuts dialog

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#399]
In Issue #320, the content, layout and design were worked out.

On review, I noticed, however (quoting from #320):

> when I open the dialog with VoiceOver, I get some unexpected verbosity (see screen shot). It is normal for a screen reader to read out the number of headings or items, but the case below, it says 3 items (which is incorrect) and then it reads out a mix of content and actual heading content.

> @jessegreenberg , what do you get with NVDA? 

> I don't remember VoiceOver being this verbose.

> <img width="915" alt="screen shot 2018-04-12 at 21 22 01" src="https://user-images.githubusercontent.com/1902116/38709925-406b0c14-3e98-11e8-9486-07ff93acf323.png">

**Question:**
I am wondering why Voice Over is saying 3 items, and reading a mixture of headings and content. Could some aspect of the HTML might be confusing VoiceOver?

This sim is in Dev Testing, so tagging phetsims/QA#112.

**Comment by jessegreenberg** [comment]
In NVDA and Firefox, NVDA just reads everything in the dialog from top to bottom automatically. That is how it has always been for NVDA. What would you expect for VoiceOver @terracoda? Have you tried VoiceOver with any other dialogs outside of PhET?

**Comment by jessegreenberg** [comment]
My guess is that VO is saying 3 items because the dialog has a close button, title, and a child with the rest of the content.

**Comment by jessegreenberg** [comment]
I am not sure what it should say. When I try the dialog on https://www.w3.org/TR/2017/NOTE-wai-aria-practices-1.1-20171214/examples/dialog-modal/dialog.html, VoiceOver just says

>"Edit text with auto fill  menu"

when I open the dialog.

**Comment by jessegreenberg** [comment]
The good news is that our dialog HTML looks very similar to the example at the above link - I think the main difference is that focus for them goes to the first form element,  focus for us goes to the "close" button (because there is no form element).

**Comment by terracoda** [comment]
I would expect VoiceOver to something sensible like read from the top, or say, dialog with x number of items where x is the correct number of actual headings and list items combined. Another sensible thing would be to just announce the number of headings.

**Comment by terracoda** [comment]
We could try placing focus on the h1, the first static element of the dialog. This is the design pattern in some cases where a placing focus on the first focusable item would cause scrolling. If this method works in all AT, then I would put the close button at the bottom again.

Quoting from the ARIA 1.1 Practices:
>If content is large enough that focusing the first interactive element could cause the beginning of content to scroll out of view, it is advisable to add tabindex="-1" to a static element at the top of the dialog, such as the dialog title or first paragraph, and initially focus that element. 

Regarding the example, there is something missing because I do **NOT** hear the input element's label text. There are no `for` attributes - that might be the reason.

Regarding, "three items" in our dialog, what's the the child that is the third item?

Do you mind pasting the html code?

**Comment by terracoda** [comment]
@jessegreenberg, I'll try another sim, too.

**Comment by terracoda** [comment]
The other sims read out something similar:
> Web dialog, 3 items, close button and then the content of the headings.

I just noticed in BASE we are missing a parent UL on the first section - that might be the issue.

If you can't have a single list item, you could use a paragraph element instead and see if that fixes the issue.

Inspector screen
<img width="651" alt="screen shot 2018-04-13 at 21 32 55" src="https://user-images.githubusercontent.com/1902116/38762083-df86bd32-3f62-11e8-9efc-d89999aaef74.png">
shot:

**Comment by jessegreenberg** [comment]
Good catch @terracoda, thanks! Lets see if making that a paragraph fixes this.

**Comment by jessegreenberg** [comment]
Change made in the above commit. But it does not fix the problem.

**Comment by jessegreenberg** [comment]
If we remove the div with id "430-1117-1122-1121-427", VO says 

>"Keyboard Shortcuts, web Dialog, with 8 items, Close Button"

So VO has decided to count the number of children under the element with role="dialog", and then maybe read some information if the number items is small enough. Here is the HTML if we do this.

```html
<div tabindex="-1" id="430-1117-1122" role="dialog" aria-labelledby="430-1117-1122-1121-1119">
    <button id="430-1117-1122-1120" tabindex="-1" class="">Close</button>
    <h1 tabindex="-1" id="430-1117-1122-1121-1119">Keyboard Shortcuts</h1>
    <h2 tabindex="-1" id="430-1117-1122-1121-427-426-425-43-39">‪Grab or Release Balloon‬</h2>
    <p tabindex="-1" id="430-1117-1122-1121-427-426-425-43-42-41-38">Grab or release the balloon with Space or Enter keys.</p>
    <h2 tabindex="-1" id="430-1117-1122-1121-427-426-425-424-420">‪Move or Jump Grabbed Balloon‬</h2>
    <ul tabindex="-1" id="430-1117-1122-1121-427-426-425-424-423-422">
        <li tabindex="-1" id="430-1117-1122-1121-427-426-425-424-423-422-226">Move grabbed balloon up, left, down, or right with Arrow keys or with letter keys W, A, S, or D.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-425-424-423-422-326-325">Move slower with shift plus Arrow keys or Shift plus letter keys W, A, S, or D.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-425-424-423-422-350">Jump close to sweater with J plus S.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-425-424-423-422-373">Jump to wall with J plus W.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-425-424-423-422-396">Jump to near wall with J plus N.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-425-424-423-422-419">Jump to center with J plus C.</li>
    </ul>
    <h2 tabindex="-1" id="430-1117-1122-1121-427-426-150-146">‪General Navigation‬</h2>
    <ul tabindex="-1" id="430-1117-1122-1121-427-426-150-149-148">
        <li tabindex="-1" id="430-1117-1122-1121-427-426-150-149-148-60">Move to next item with Tab key.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-150-149-148-89">Move to previous item with Shift plus Tab key.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-150-149-148-145">Move between items in a group with Left and Right arrow keys or Up and Down Arrow keys.</li>
        <li tabindex="-1" id="430-1117-1122-1121-427-426-150-149-148-106">Exit a dialog with Escape key.</li>
    </ul>
</div>
```

@terracoda is that better?

**Comment by terracoda** [comment]
I was going to suggest removing the content wrapper div, so good idea! 

I think "8 items" is more accurate for sure, and gives an indication that there is some substantial stuff in there to read.

**Comment by terracoda** [comment]
Let's roll with this, and see, since it sounds fine with JAWS and NVDA!

**Comment by jessegreenberg** [comment]
Sounds great, thanks! Seems pretty safe but Ill just double check on NVDA and JAWS.

**Comment by terracoda** [comment]
Note that we have learned that div can effect how AT works with the content, they are not completely ignored!

**Comment by jessegreenberg** [comment]
Indeed @terracoda, we will have to be on the lookout for this!

**Comment by jessegreenberg** [comment]
Ah, the div was there before because the recommendation (at least at the time) was to use the `aria-describedby` attribute on a parent containing the dialog content without the close button and the title.

It seems like we have to choose between which AT we want it to sound best on. I don't much care either way. @terracoda do you think we should keep the div or no?

**Comment by jessegreenberg** [comment]
With the div removed, JAWS doesn't read the dialog content automatically.

**Comment by terracoda** [comment]
@jessegreenberg, I'll have another listen!

**Comment by terracoda** [comment]
@jessegreenberg, re https://github.com/phetsims/balloons-and-static-electricity/issues/399#issuecomment-381292664, you also removed the `aria-describedby`.

Did you do that intentionally? Did you hear a difference?

I'm reviewed some web articles and examples:
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_dialog_role
- https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html

This article outlines the design pattern nicely, but not with code
- https://heydesigner.com/blog/best-practices-modals-overlays-dialog-windows/

**Comment by terracoda** [comment]
@jessegreenberg, in the w3c example, Voice Over is not recognizing the aria-labelledby attribute which is strange, but in our case VoiceOver is naming the dialog.

Could you temporarily remove the `arai-describedby` attribute and I can see if it sounds any better without that. 

My reasoning:
- The h1, "Keyboard Shortcuts" of the dialog sufficiently describe the purpose of and what is inside the dialog. 
- The `aria-describedby` attribute is optional, and perhaps is not adding any benefit in our case.

**Comment by jessegreenberg** [comment]
> @jessegreenberg, re #399 (comment), you also removed the aria-describedby.

Yes, without the div what would aria-describedby point to?

To clarify, you want to try including the `div`, without the `aria-describedby` attribute?

**Comment by terracoda** [comment]
@jessegreenberg, yes, I think we should try the dialog with the main `div`, but without the `aria-describedby` attribute.

**Comment by terracoda** [comment]
That is, try it as-is, but without `aria-describedby`.

**Comment by jessegreenberg** [comment]
Sounds good @terracoda I remove the aria-describedy but kept the div, should be ready on phettest after pulling.

In NVDA, it sounds like:
> Keyboard Shortcuts dialog
clickable Close button
Keyboard Shortcuts
button Close

Then the rest of the dialog content is read in full.

**Comment by terracoda** [comment]
Thanks @jessegreenberg, I'll test in VoiceOver.

**Comment by terracoda** [comment]
@jessegreenberg, much better! Thanks for taking the time to help me try all these things!
We need to document this our eventual design patterns!
<img width="652" alt="image" src="https://user-images.githubusercontent.com/1902116/38947163-48e83c40-4316-11e8-97ba-f32ec8e8c4a8.png">

**Comment by terracoda** [comment]
@jessegreenberg, also, after I hear the initial dialog information, I just use the cursor keys to key on, and it sounds good.

**Comment by terracoda** [comment]
We might need to keep our eyes on the HTML5 `dialog` element for the future, but for now I think we have a dialog that works.

Please close this issue when you are ready!

**Comment by terracoda** [comment]
@jessegreenberg, regarding https://github.com/phetsims/balloons-and-static-electricity/issues/399#issuecomment-382235184

As long as the dialog is announced properly in JAWS and the cursor keys allow the user to read the content in the dialog in the correct order, the dialog is good from my perspective. 

The content in the dialog does not have to be read out automatically to be accessible.

**Comment by jessegreenberg** [comment]
> We might need to keep our eyes on the HTML5 dialog element for the future, but for now I think we have a dialog that works.

Sounds great.

> As long as the dialog is announced properly in JAWS and the cursor keys allow the user to read the content in the dialog in the correct order, the dialog is good from my perspective.

Understood, thanks for clarifying! It sounds good in JAWS so I think this issue can be closed. Thanks  @terracoda.
